### PR TITLE
File upload error message templates

### PR DESCRIPTION
**Pull request by stevenaproctor**
Worked with @Emilyallinson to come up with standard error message templates for file upload. Reviewed with @adamsilver to make sure it met HMCTS needs.

### CODE REVIEW

File: src/components/file-upload/index.md.njk [hunk 1]
```diff
@@ -23,6 +23,54 @@ There are 2 ways to use the file upload component. You can use HTML or, if you
 
 {{ example({group: "components", item: "file-upload", example: "default", html: true, nunjucks: true, open: false, id: "default-2"}) }}
 
+### Error messages
+
+Check that a valid date is entered by the user. Invalid dates should be reported to the user like this:
```
Comment (36degrees) [inline src/components/file-upload/index.md.njk]: These error messages don't relate to dates.

File: src/components/file-upload/index.md.njk [hunk 2]
```diff
@@ -23,6 +23,54 @@ There are 2 ways to use the file upload component. You can use HTML or, if you
...
+
+Make sure errors follow the guidance in [error message](/components/error-message/) and have specific error messages for specific error states.
+
+#### If nothing has been uploaded
+Say ‘Select a [whatever they need to upload]’.<br>
+For example, ‘Select a report’.
+
+#### If the file is the wrong file type
+Say‘The [whatever] must be a [list of file types]’.<br>
+For example, ‘The report must be a CSV or ODS’ or ‘The upload must be a JPG, BMP, PNG, TIF or PDF’.
```
Comment (36degrees) [inline src/components/file-upload/index.md.njk]: I'm not sure how well 'upload' works as a noun here – how about

```suggestion
For example, ‘The report must be a CSV or ODS’ or ‘The uploaded file must be a JPG, BMP, PNG, TIF or PDF’.
```

?

File: src/components/file-upload/index.md.njk [hunk 3]
```diff
@@ -23,6 +23,54 @@ There are 2 ways to use the file upload component. You can use HTML or, if you
...
+
+#### If the file is too big
+Say ‘The [whatever] must be smaller than [largest file size]’.<br>
+For example, ‘The CSV must be smaller than 2MB’.
+
+#### If the file is empty
+Say ‘The [whatever] is empty – select a [whatever] with data in it’.<br>
+For example, ‘The CSV is empty – select a CSV with data in it’.
```
Comment (36degrees) [inline src/components/file-upload/index.md.njk]: We're alternating between the concept of 'uploading' and the concept of 'selecting' – can we try and stick to one?

File: src/components/file-upload/index.md.njk [hunk 4]
```diff
@@ -23,6 +23,54 @@ There are 2 ways to use the file upload component. You can use HTML or, if you
...
+
+#### If there was a problem and the file was not uploaded
+Say ‘The [whatever] could not be uploaded – try again’.<br>
+For example, ‘The CSV could not be uploaded – try again’.
+
+#### If the user can upload more than one file but there is a limit on how many they can upload
+Say ‘You can only upload up to [highest number] [whatevers] at the same time’.<br>
+For example, ‘You can only upload up to 10 reports at the same time’.
+
+#### If the upload must be in a template that has been changed and you do know what has changed
```
Comment (36degrees) [inline src/components/file-upload/index.md.njk]: I find this heading (and the next one) really hard to parse. I think it's mainly because of the use of 'upload' as a noun, coupled with the phrase 'in a template' which I don't think really makes sense?

File: src/components/file-upload/index.md.njk (hunk 2 above)
Comment (36degrees) [reply src/components/file-upload/index.md.njk]: Actually, Would using 'selected file' work better throughout?

- ‘The selected file must be a CSV or ODS’ or ‘The selected file must be a JPG, BMP, PNG, TIF or PDF’.
- ‘The selected file must be smaller than 2MB’
- ‘The selected file is empty – select a CSV with data in it’
- ‘The selected file contains a virus’
- ‘The selected file is password protected’

File: src/components/file-upload/index.md.njk (hunk 3 above)
Comment (36degrees) [reply src/components/file-upload/index.md.njk]: Also not sure that 'a CSV with data in it' is a good description of what's expected.

File: src/components/file-upload/index.md.njk (hunk 1 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: Good spot. I have changed this to match other documentation in the design system.

File: src/components/file-upload/index.md.njk (hunk 2 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: 'uploaded file' is better than 'upload' and I will stick with upload rather than select.

File: src/components/file-upload/index.md.njk (hunk 3 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: I think this is because of the different terminology used in the two different services. I will stick to upload because that is what is in the standard example.

Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: The empty file error is odd but I think 'a CSV with data in it' is all we can say without any more context. Do you think 'The [whatever] is empty' without the extra information is better?

File: src/components/file-upload/index.md.njk (hunk 4 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: I agree. I have tried to make them better.

File: src/components/file-upload/index.md.njk (hunk 2 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: Using 'selected file' for all messages is a good idea. It means they will be less specific than using the file type or description of what the file is.

Maybe a couple of examples would make sense with selected file and with something more specific. I have added this to the latest commit.

Comment (amyhupe) [reply src/components/file-upload/index.md.njk]: If possible, can we just exclusively use selected file? I think it works fine, and encouraging people to switch between naming the thing (eg, 'the report') and using 'selected file' seems like it could be problematic.

File: src/components/file-upload/index.md.njk [hunk 5]
```diff
@@ -23,6 +23,54 @@ There are 2 ways to use the file upload component. You can use HTML or, if you
...
+
+#### If the file is password protected
+Say ‘The [whatever] is password protected’.<br>
+For example, ‘The report is password protected’ or ‘The selected file is password protected’.
+
+#### If there was a problem and the file was not uploaded
+Say ‘The [whatever] could not be uploaded – try again’.<br>
+For example, ‘The CSV could not be uploaded – try again’ or ‘The selected file could not be uploaded – try again’.
+
+#### If the user can select more than one file but there is a limit on how many they can select
```
Comment (amyhupe) [inline src/components/file-upload/index.md.njk]: Do we need to say 'If the user can select more than one file'? Couldn't we simply say 'If there is a limit on how many files a user can select' to make it more concise?

File: src/components/file-upload/index.md.njk (hunk 4 above)
Comment (amyhupe) [reply src/components/file-upload/index.md.njk]: I'm still finding these quite clunky and difficult to read. I'm also just wondering if this level of specificity is necessary? How common a use case actually is this? If we were to hold this up to the Design System criteria – is there evidence that multiple services will be needing (and able) to scan uploaded files to check for changes? I'd honestly be tempted to leave these out – what do you think @stevenaproctor?

File: src/components/file-upload/index.md.njk (hunk 2 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: Agreed.

File: src/components/file-upload/index.md.njk (hunk 3 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: I agree it is specific enough. Going with 'The selected file is empty' to be consistent.

File: src/components/file-upload/index.md.njk (hunk 5 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: Agreed.

File: src/components/file-upload/index.md.njk (hunk 4 above)
Comment (stevenaproctor) [reply src/components/file-upload/index.md.njk]: I know of at least two services in HMRC that has templates that people must use for uploads. But the templates are very different so these kinds of errors are had to create templates for.

I will remove these two and leave the general template one at the bottom as a starter.

**Comment by stevenaproctor** [comment]
@amyhupe I have made changes that I think cover all your points. I have purposely left the final error in as a general error and indication it is something people may have to have an error for. Let me know what you think.

**Comment by amyhupe** [comment]
@stevenaproctor - I'm happy with this now, thanks for making those changes. I've just corrected a couple of small typos.

I'm also going to do a separate PR once we've merged this to add a link to these messages from the error message page, as we do with the others. 

Thanks

**Comment by stevenaproctor** [comment]
@amyhupe Thanks for fixing the typos and sorting. Updating the error message page is part of this PR so that should be already done once it has merged.

**Comment by amyhupe** [comment]
That's awkward! Totally missed it, sorry. Will merge now.
